contracts/ForeignBridge.sol

```diff
@@ -2,74 +2,92 @@ pragma solidity 0.4.19;
+    uint256 homeGasPrice = 1000000000 wei;
+    mapping (bytes32 => bytes) messages;
+    mapping (bytes32 => bytes) signatures;
+    mapping (bytes32 => bool) messages_signed;
+    mapping (bytes32 => uint) num_messages_signed;
     mapping (bytes32 => bool) deposits_signed;
-    mapping (bytes32 => uint8) num_deposits_signed;
-	mapping (bytes32 => bool) mintRequestsDone;
-	mapping (bytes32 => uint8) withdrawRequests;
+    mapping (bytes32 => uint) num_deposits_signed;
 
-    /// Pending signatures and authorities who confirmed them
-    BridgeValidators public validatorsContract;
+    mapping (bytes32 => bool) tokenAddressApprovalSigns;
+    mapping (address => uint256) public numTokenAddressApprovalSigns;
 
-    /// Token contract which the bridge has full ownership to burn and mint
-    address public token;
+    IBurnableMintableERC677Token public erc677token;
 
     /// triggered when relay of deposit from HomeBridge is complete
     event Deposit(address recipient, uint value);
 
     /// Event created on money withdraw.
-    event Withdraw(address recipient, uint value);
+    event Withdraw(address recipient, uint256 value, uint256 homeGasPrice);
 
     /// Collected signatures which should be relayed to home chain.
     event CollectedSignatures(address authorityResponsibleForRelay, bytes32 messageHash);
 
+    event TokenAddress(address token);
+
+    event GasConsumptionLimitsUpdated(uint256 gasLimitDepositRelay, uint256 gasLimitWithdrawConfirm);
+
+    event SignedForDeposit(address indexed signer, bytes32 message);
+    event SignedForWithdraw(address indexed signer, bytes32 message);
+
     function ForeignBridge(
-        address _token,
-        address _validatorsContract
-    ) public
-    {
-        require(_token != address(0) && _validatorsContract != address(0));
-        validatorsContract = BridgeValidators(_validatorsContract);
-        token = _token;
+        address _validatorContract,
+        address _erc677token
+    ) public Validatable(_validatorContract) {
+        erc677token = IBurnableMintableERC677Token(_erc677token);
     }
 
+    function setGasLimitDepositRelay(uint256 gas) onlyValidator {
+        gasLimitDepositRelay = gas;
 
-    /// Used to deposit money to the contract.
-    ///
-    /// deposit recipient (bytes20)
-    /// deposit value (uint)
-    /// mainnet transaction hash (bytes32) // to avoid transaction duplication
-    function deposit(address recipient, uint256 value, bytes32 transactionHash) public {
-        require(validatorsContract.onlyValidator(msg.sender));
+        GasConsumptionLimitsUpdated(gasLimitDepositRelay, gasLimitWithdrawConfirm);
+    }
+
+    function setGasLimitWithdrawConfirm(uint256 gas) onlyValidator {
+        gasLimitWithdrawConfirm = gas;
+
+        GasConsumptionLimitsUpdated(gasLimitDepositRelay, gasLimitWithdrawConfirm);
+    }
+
+    function deposit(address recipient, uint value, bytes32 transactionHash) public onlyValidator {
```

Comment:
To not forget: create an issue to implement functionality that the bridge calls `isValidator` before sending a transaction.

Reply:
???

Reply:
https://github.com/poanetwork/parity-bridge/issues/28